File: src/matlab/com/springbok/pattern/PatternSRR_405V01.m
classdef PatternSRR_405V01 < SpacePattern & TransmitPattern
% Describes the ITU antenna pattern SRR_405V01.
  
% Copyright (C) 2016 Springbok LLC <http://springbok.io/> All rights reserved.

  properties (SetAccess = private, GetAccess = public)
    
    % Maximum antenna gain [dB]
    GainMax
    % Cross-sectional half-power beamwidth, degrees
    Phi0
    
    % Intermediate gain calculation results
    
  end % properties (SetAccess = private, GetAccess = public)
  
  methods
    
    function this = PatternSRR_405V01(Phi0)
    % Constructs a PatternSRR_405V01 given a cross-sectional half-power beamwidth.
    %
    % Parameters
    %   GainMax - Maximum antenna gain [dB]
    %   Phi0 - Cross-sectional half-power beamwidth, degrees
      
    % No argument constructor
      if nargin == 0
        return;
        
      end % if
      
      % Check number and class of input arguments.
      if ~(nargin == 1 ...
           && isnumeric(Phi0))
        exc = SException( ...
            'Springbok:InvalidInput', ...
            'Numeric Phi0 must be provided.');
        throw(exc)
        
      end % if
      
      % Validate input parameters.
      PatternUtility.validate_input('Phi0', Phi0);
      
      % Assign properties
      this.Phi0 = Phi0;
      
      % Compute derived properties
      % None
      
    end % PatternSRR_405V01()
    
    function that = copy(this)
    % Copies a PatternSRR_405V01 given a maximum antenna gain.
    %
    % Parameters
    %   None
      that = PatternSRR_405V01(this.Phi0);

    end % copy()

    function [G, Gx, fH] = gain(this, Phi, varargin)
    % Appendix 30 (RR-2001) reference transmitting space station
    % antenna pattern for Regions 1 and 3.
    %
    % Parameters:
    %   Phi - Angle for which a gain is calculated [degrees]
    %
    % Optional parameters (entered as key/value pairs):
    %   PlotFlag - Flag to indicate plot, true or {false}
    %
    % Returns:
    %   G - Co-polar gain [dB]
    %   Gx - Cross-polar gain [dB]
    %   fH - figure handle
      
    % Check number and class of input arguments.
      if ~(nargin >= 1 ...
           && isnumeric(Phi))
        exc = SException( ...
            'Springbok:InvalidInput', ...
            'Numeric Phi must be provided.');
        throw(exc)
        
      end % if
      
      % Validate input parameters.
      PatternUtility.validate_input('Phi', Phi);
      
      % Handle optional inputs.
      input = struct();
      if ~isempty(varargin)
        nVarargin = size(varargin, 2);
        
        if rem(nVarargin, 2)
          exc = SException( ...
              'Springbok:InvalidInput', ...
              'Inputs must be key/value pairs.');
          throw(exc)
          
        else
          for iVarargin = 1:2:nVarargin
            key = varargin{iVarargin};
            value = varargin{iVarargin + 1};
            input = PatternUtility.validate_input(key, value, input);
            
          end % for
          
        end % if
        
      end % if
      
      % TODO: Test constructor optional agruments
      % Assign default values to any optional input parameters that
      % were not included in varargin.
      if ~isfield(input, 'GainMax')
        this.GainMax = 0;
        absolute_pattern = false;
        
      else
        this.GainMax = input.GainMax;
        absolute_pattern = true;
        
      end % if
      if ~isfield(input, 'PlotFlag')
        PlotFlag = false;

      else
        PlotFlag = input.PlotFlag;

      end % if
      if ~isfield(input, 'DoValidate')
        DoValidate = true;

      else
        DoValidate = input.DoValidate;

      end % if

      % Implement pattern.
      phi = max(eps, Phi); 
      phi_over_phi0 = phi ./ this.Phi0;
      
      G0 = this.GainMax - 12 * phi_over_phi0 .^ 2;
      G1 = this.GainMax - 30;
      G2 = this.GainMax - 17.5 - 25 * log10(phi_over_phi0);
      
      G = G0 .* (0    <= phi_over_phi0 & phi_over_phi0 <= 1.58) ...
          + G1 .* (1.58 <  phi_over_phi0 & phi_over_phi0 <= 3.16) ...
          + G2 .* (3.16 <  phi_over_phi0);
      
      phix = max(eps, abs(phi_over_phi0 - 1));
      Gx0 = this.GainMax - 40 - 40 * log10(phix);
      Gx1 = this.GainMax - 33;
      Gx2 = this.GainMax - 40 - 40 * log10(phix);
      
      Gx = Gx0 .* (0    <= phi_over_phi0 & phi_over_phi0 <= 0.33) ...
           + Gx1 .* (0.33 <  phi_over_phi0 & phi_over_phi0 <= 1.67) ...
           + Gx2 .* (1.67 <  phi_over_phi0);
      
      % Apply "flooring" for absolute gain pattern.
      if absolute_pattern
        G = max(0, G);
        Gx = max(0, Gx);
        
      end % if
      
      % Validate low level rules.
      if this.GainMax < 33 && absolute_pattern
        warning( ...
            'Springbok:InvalidResult', ...
            'GainMax is less than 33.');
        
      end % if
      
      % Validate output parameters.
      if DoValidate
        PatternUtility.validate_output(G, Gx, this.GainMax);
      
      end % if
      
      % Plot results, if requested.
      if PlotFlag == true
        fH = figure;
        semilogx(phi, G, 'k')
        grid on
        hold on
        semilogx(phi, Gx, 'k')
        v = axis;
        axis([0.1 180 v(3) v(4)])
        
        if absolute_pattern
          ttlStr = [strrep(mfilename, '_', '\_'), ' Absolute Reference Antenna Pattern'];
          
        else
          ttlStr = [strrep(mfilename, '_', '\_'), ' Relative Reference Antenna Pattern'];
          
        end % if
        
        title(ttlStr)
        xlabel('Phi (degrees)')
        ylabel('Gain (dB)')
        
      else
        fH = [];
        
      end % if
      
    end % gain()
    
  end % methods
  
end % classdef

File: src/java/com/springbok/pattern/PatternSRR_405V01.java
/* Copyright (C) 2022 Springbok LLC

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or (at
your option) any later version.

This program is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/
package com.springbok.pattern;

import com.springbok.utility.PatternUtility;

import java.util.Map;

/**
 * Describes the ITU antenna pattern SRR_405V01.
 */
public class PatternSRR_405V01 implements SpacePattern, TransmitPattern {

	/**
	 * Maximum antenna gain [dB]
	 */
	private double GainMax;

	/**
	 * Cross-sectional half-power beamwidth [degrees]
	 */
	private double Phi0;

	/**
	 * Constructs a PatternSRR_405V01 given a cross-sectional half-power beamwidth.
	 *
	 * @param Phi0    Cross-sectional half-power beamwidth [degrees]
	 */
	public PatternSRR_405V01(double Phi0) {
		/* Validate input parameters */
		PatternUtility.validate_input("Phi0", Phi0);

		/* Assign properties */
		this.Phi0 = Phi0;

		/* Compute derived properties */
		/* None */
	}

	/**
	 * Gets maximum antenna gain [dB].
	 *
	 * @return Maximum antenna gain [dB]
	 */
	public double getGainMax() {
		return GainMax;
	}

	/**
	 * Gets Cross-sectional half-power beamwidth [degrees].
	 *
	 * @return Cross-sectional half-power beamwidth [degrees]
	 */
	public double getPhi0() {
		return Phi0;
	}

	/**
	 * Copies a PatternSRR_405V01 given a maximum antenna gain.
	 */
	public PatternSRR_405V01 copy() {
		return new PatternSRR_405V01(this.Phi0);
	}

	/**
	 * Appendix 30 (RR-2001) reference transmitting space station antenna pattern
	 * for Regions 1 and 3.
	 *
	 * @param Phi     Angle for which a gain is calculated [degrees]
	 * @param options Optional parameters (entered as key/value pairs)
	 * @return Co-polar and cross-polar gain [dB]
	 */
	public Gain gain(double Phi, Map options) {
		/* Validate input parameters */
		PatternUtility.validate_input("Phi", Phi);

		/* Validate input options */
		options = PatternUtility.validateOptions(options);

		// TODO: Test constructor optional agruments
		GainMax = 0;
		boolean absolute_pattern = false;
		if (options.containsKey("GainMax")) {
			GainMax = (double) options.get("GainMax");
			absolute_pattern = true;
		}

		/* Implement pattern */
		double eps = Math.ulp(1.0);
		double phi = Math.max(eps, Phi);

		double phi_over_phi0 = phi / Phi0;

		double G0 = GainMax - 12 * Math.pow(phi_over_phi0, 2);
		double G1 = GainMax - 30;
		double G2 = GainMax - 17.5 - 25 * Math.log10(phi_over_phi0);

		double G = G0 * (0 <= phi_over_phi0 && phi_over_phi0 <= 1.58 ? 1 : 0)
				+ G1 * (1.58 < phi_over_phi0 && phi_over_phi0 <= 3.16 ? 1 : 0)
				+ G2 * (3.16 < phi_over_phi0 ? 1 : 0);

		double phix = Math.max(eps, Math.abs(phi_over_phi0 - 1));
		double Gx0 = GainMax - 40 - 40 * Math.log10(phix);
		double Gx1 = GainMax - 33;
		double Gx2 = GainMax - 40 - 40 * Math.log10(phix);

		double Gx = Gx0 * (0 <= phi_over_phi0 && phi_over_phi0 <= 0.33 ? 1 : 0)
				+ Gx1 * (0.33 < phi_over_phi0 && phi_over_phi0 <= 1.67 ? 1 : 0)
				+ Gx2 * (1.67 < phi_over_phi0 ? 1 : 0);

		/* Apply "flooring" for absolute gain pattern */
		if (absolute_pattern) {
			G = Math.max(0, G);
			Gx = Math.max(0, Gx);
		}
		/* Validate low level rules. */
		if (GainMax < 33 && absolute_pattern) {
			PatternUtility.logger.warn("GainMax is less than 33.");
		}
		/* Validate output parameters */
		if ((boolean) options.get("DoValidate")) {
			PatternUtility.validate_output(G, Gx, GainMax);
		}
		return new Gain(G, Gx);
	}

	/*
	 * (non-Javadoc)
	 *
	 * @see java.lang.Object#hashCode()
	 */
	@Override
	public int hashCode() {
		final int prime = 31;
		int result = 1;
		long temp;
		temp = Double.doubleToLongBits(Phi0);
		result = prime * result + (int) (temp ^ (temp >>> 32));
		temp = Double.doubleToLongBits(GainMax);
		result = prime * result + (int) (temp ^ (temp >>> 32));
		return result;
	}

	/*
	 * (non-Javadoc)
	 *
	 * @see java.lang.Object#equals(java.lang.Object)
	 */
	@Override
	public boolean equals(Object obj) {
		if (this == obj) {
			return true;
		}
		if (obj == null) {
			return false;
		}
		if (!(obj instanceof PatternSRR_405V01)) {
			return false;
		}
		PatternSRR_405V01 other = (PatternSRR_405V01) obj;
		if (Double.doubleToLongBits(Phi0) != Double.doubleToLongBits(other.Phi0)) {
			return false;
		}
		if (Double.doubleToLongBits(GainMax) != Double.doubleToLongBits(other.GainMax)) {
			return false;
		}
		return true;
	}
}
